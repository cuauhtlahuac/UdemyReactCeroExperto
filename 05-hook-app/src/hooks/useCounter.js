import { useState } from "react"

export default function useCounter( initialState = 13 ) {
    const [counter, setCounter] = useState( initialState )

    function decrement(factor = 1) {
        setCounter(( counter <= 0 ) ? 0 : counter - factor);
    };

    function reset() {
        setCounter( initialState );
    };

    function increment(factor = 1) {
        setCounter(counter + factor);
    };

    return {
        counter,
        decrement,
        reset,
        increment,
    }
}
