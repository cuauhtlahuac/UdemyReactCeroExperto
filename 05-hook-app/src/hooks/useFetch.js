import { useState, useEffect, useRef } from 'react';

export default function useFetch( url ) {

	const initialState = {
		data: null,
		loading: true,
		error: null,
	};

	const [ state, setState ] = useState( initialState );

	const isMounted = useRef(true);

	useEffect(() => {
		
		return () => {
			isMounted.current = false;
		}
	}, [isMounted])

	useEffect(() => {
		setState({
			data: null,
			loading: true,
			error: null,
		});

        fetch( url )
        .then( resp => resp.json() )
        .then( data => { 

			if(isMounted.current){
				setState({
					data,
					loading: false,
					error: null,
				})
			};
		}).catch( error => (setState({
			data: null,
			loading: false,
			error: error,
		})))
	}, [ url ]);

	return state;
}
