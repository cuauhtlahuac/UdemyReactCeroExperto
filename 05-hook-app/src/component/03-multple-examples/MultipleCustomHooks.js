import React from 'react'
import useFetch from 'hooks/useFetch'
import useCounter from 'hooks/useCounter'

const MultipleCustomHooks = () => {

    const { counter, increment } = useCounter(1)

    const url = `https://www.breakingbadapi.com/api/quotes/${ counter }`

    const {
        data,
        loading,
    } = useFetch( url );

    const { author, quote } = !!data && data[0];

    return (
        <>
          <h1>Breaking Bad Quotes</h1>
            <hr/>

            { loading ?
            (
                <div className="d-flex mb-2 justify-content-end">
                    <div className="spinner-grow bg-info" style={{ width: '3rem', height: '3rem' }} role="status">
                        <span className="sr-only">Loading...</span>
                    </div>
                </div>
            ) : (
                <blockquote className="blockquote text-right">
                    <p className="mb-0">{ quote }</p>
                    <footer className="blockquote-footer">{ author }</footer>
                </blockquote>
            )}

            <button className="btn btn-primary" onClick={ increment }>
                next quote 
            </button>
        </>
    )
}

export default MultipleCustomHooks
