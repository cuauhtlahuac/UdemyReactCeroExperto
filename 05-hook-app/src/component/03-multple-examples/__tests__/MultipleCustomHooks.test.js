import React from 'react';
import '@testing-library/jest-dom';
import { mount, shallow } from 'enzyme';

import MultipleCustomHooks from '../MultipleCustomHooks';
import useFetch from 'hooks/useFetch';
import useCounter from 'hooks/useCounter';

jest.mock('hooks/useFetch');
jest.mock('hooks/useCounter');

describe('MultipleCustomHooks Tests', () => {
  let component;

	beforeEach(() => {
    useCounter.mockReturnValue({
      counter: 21,
      increment: jest.fn(),
    })

    useFetch.mockReturnValue({
			data: null,
			loading: true,
			error: null,
    });
    
		component = mount(<MultipleCustomHooks />);
	});
  
	afterEach(() => {
    component.unmount();
	});
  
	test('should render correctly', () => {
    expect(component).toMatchSnapshot();
  });
  
	test('should paint the information', () => {
    useFetch.mockReturnValue({
      data: [{
        author: 'Cuauhtlahuac',
        quote: 'Hola Mundo'
      }],
			loading: false,
			error: null,
    });

    const component = mount(<MultipleCustomHooks />);
    
    const quote = component.find('p')
    const author = component.find('footer')
    const loading = component.find('span .sr-only');

    expect(loading.exists()).toBe(false);
		expect(quote.text()).toBe('Hola Mundo');
		expect(author.text()).toBe('Cuauhtlahuac');
	});
});
