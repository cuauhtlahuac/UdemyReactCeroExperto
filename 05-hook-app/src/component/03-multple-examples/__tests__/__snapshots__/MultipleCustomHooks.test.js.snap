// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultipleCustomHooks Tests should render correctly 1`] = `
Array [
  <h1>
    Breaking Bad Quotes
  </h1>,
  <hr />,
  <div
    className="d-flex mb-2 justify-content-end"
  >
    <div
      className="spinner-grow bg-info"
      role="status"
      style={
        Object {
          "height": "3rem",
          "width": "3rem",
        }
      }
    >
      <span
        className="sr-only"
      >
        Loading...
      </span>
    </div>
  </div>,
  <button
    className="btn btn-primary"
    onClick={[MockFunction]}
  >
    next quote
  </button>,
]
`;
